builtin/logical/database/backend.go

```diff
@@ -286,3 +485,327 @@ cassandra, mssql, mysql, postgres
 the "database/config/" path.
 `
+
+// populateQueue loads the priority queue with existing static accounts. This
+// occurs at initialization, after any WAL entries of failed or interrupted
+// rotations have been processed. It lists the roles from storage and searches
+// for any that have an associated static account, then adds them to the
+// priority queue for rotations.
+func (b *databaseBackend) populateQueue(ctx context.Context, s logical.Storage) {
+	log := b.Logger()
+	log.Info("populating role rotation queue")
+
+	roles, err := s.List(ctx, "static-role/")
+	if err != nil {
+		log.Warn("unable to list role for enqueueing", "error", err)
+		return
+	}
+
+	for _, roleName := range roles {
+		select {
+		case <-ctx.Done():
+			log.Info("rotation queue restore cancelled")
+			return
+		default:
+		}
+
+		role, err := b.StaticRole(ctx, s, roleName)
+		if err != nil {
+			log.Warn("unable to read static role", "error", err, "role", roleName)
+			continue
+		}
+		if role == nil || role.StaticAccount == nil {
+			continue
+		}
+
+		if err := b.pushItem(&queue.Item{
+			Key:      roleName,
+			Priority: role.StaticAccount.LastVaultRotation.Add(role.StaticAccount.RotationPeriod).Unix(),
+		}); err != nil {
+			log.Warn("unable to enqueue item", "error", err, "role", roleName)
+		}
+	}
+}
+
+// runTicker kicks off a periodic ticker that invoke the automatic credential
+// rotation method at a determined interval. The default interval is 5 seconds.
+func (b *databaseBackend) runTicker(ctx context.Context, s logical.Storage) {
+	b.logger.Info("starting periodic ticker")
+	tick := time.NewTicker(queueTickInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			b.rotateCredentials(ctx, s)
+
+		case <-ctx.Done():
+			b.logger.Info("stopping periodic ticker")
+			return
+		}
+	}
+}
+
+// setCredentialsWAL is used to store information in a WAL that can retry a
+// credential setting or rotation in the event of partial failure.
+type setCredentialsWAL struct {
+	NewPassword string
+	OldPassword string
+	RoleName    string
+	Username    string
+
+	LastVaultRotation time.Time
+	Statements        dbplugin.Statements
+}
+
+// rotateCredentials sets a new password for a static account. This method is
+// invoked in the runTicker method, which is in it's own go-routine, and invoked
+// periodically (approximately every 5 seconds).
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		item, err := b.credRotationQueue.Pop()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn(fmt.Sprintf("unable load role (%s)", item.Key), "error", err)
```

Comment:
I think we don't want to use Sprintf for log messages.  The role should become a key in this warning message.

Reply:
Done in 236dc1868